File: apps/frontend/src/shared/api/emailChange.ts
import { apiClient } from './client';

export interface RequestEmailChangeRequest {
  new_email: string;
}

export interface RequestEmailChangeResponse {
  request_id: string;
  message: string;
}

export interface VerifyEmailRequest {
  request_id: string;
  verification_code: string;
}

export interface VerifyEmailResponse {
  message: string;
  status: string;
}

export interface CompleteEmailChangeRequest {
  request_id: string;
}

export interface CompleteEmailChangeResponse {
  message: string;
}

export interface CancelEmailChangeRequest {
  request_id: string;
}

export interface CancelEmailChangeResponse {
  message: string;
}

export interface ResendVerificationCodeRequest {
  request_id: string;
}

export interface ResendVerificationCodeResponse {
  message: string;
}

export interface EmailChangeRequestResponse {
  id: string;
  old_email: string;
  new_email: string;
  status: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
}

export interface EmailChangeStatusResponse {
  has_active_request: boolean;
  request?: EmailChangeRequestResponse;
}

export interface EmailChangeAuditLogResponse {
  id: string;
  old_email: string;
  new_email: string;
  change_type: string;
  ip_address?: string;
  user_agent?: string;
  created_at: string;
}

export interface EmailChangeAuditLogsResponse {
  logs: EmailChangeAuditLogResponse[];
  total: number;
}

export const emailChangeApi = {
  // Создать запрос на смену email
  requestEmailChange: async (data: RequestEmailChangeRequest): Promise<RequestEmailChangeResponse> => {
    const response = await apiClient.post('/email-change/request', data);
    return response.data.data;
  },

  // Подтвердить старый email
  verifyOldEmail: async (data: VerifyEmailRequest): Promise<VerifyEmailResponse> => {
    const response = await apiClient.post('/email-change/verify-old', data);
    return response.data.data;
  },

  // Подтвердить новый email
  verifyNewEmail: async (data: VerifyEmailRequest): Promise<VerifyEmailResponse> => {
    const response = await apiClient.post('/email-change/verify-new', data);
    return response.data.data;
  },

  // Завершить смену email
  completeEmailChange: async (data: CompleteEmailChangeRequest): Promise<CompleteEmailChangeResponse> => {
    const response = await apiClient.post('/email-change/complete', data);
    return response.data.data;
  },

  // Отменить запрос на смену email
  cancelEmailChange: async (data: CancelEmailChangeRequest): Promise<CancelEmailChangeResponse> => {
    const response = await apiClient.post('/email-change/cancel', data);
    return response.data.data;
  },

  // Повторно отправить код подтверждения
  resendVerificationCode: async (data: ResendVerificationCodeRequest): Promise<ResendVerificationCodeResponse> => {
    const response = await apiClient.post('/email-change/resend', data);
    return response.data.data;
  },

  // Получить статус активного запроса
  getEmailChangeStatus: async (): Promise<EmailChangeStatusResponse> => {
    const response = await apiClient.get('/email-change/status');
    return response.data.data;
  },

  // Получить аудит-лог изменений email
  getAuditLogs: async (limit = 20, offset = 0): Promise<EmailChangeAuditLogsResponse> => {
    const response = await apiClient.get(`/email-change/audit-logs?limit=${limit}&offset=${offset}`);
    return response.data.data;
  },
};

File: apps/backend/internal/dto/email_change.go
package dto

import "time"

// RequestEmailChangeRequest запрос на создание запроса смены email
type RequestEmailChangeRequest struct {
	NewEmail string `json:"new_email" validate:"required,email"`
}

// RequestEmailChangeResponse ответ на создание запроса смены email
type RequestEmailChangeResponse struct {
	RequestID string `json:"request_id"`
	Message   string `json:"message"`
}

// VerifyEmailRequest запрос на подтверждение email
type VerifyEmailRequest struct {
	RequestID        string `json:"request_id" validate:"required"`
	VerificationCode string `json:"verification_code" validate:"required,len=6"`
}

// VerifyEmailResponse ответ на подтверждение email
type VerifyEmailResponse struct {
	Message string `json:"message"`
	Status  string `json:"status"`
}

// CompleteEmailChangeRequest запрос на завершение смены email
type CompleteEmailChangeRequest struct {
	RequestID string `json:"request_id" validate:"required"`
}

// CompleteEmailChangeResponse ответ на завершение смены email
type CompleteEmailChangeResponse struct {
	Message string `json:"message"`
}

// CancelEmailChangeRequest запрос на отмену смены email
type CancelEmailChangeRequest struct {
	RequestID string `json:"request_id" validate:"required"`
}

// CancelEmailChangeResponse ответ на отмену смены email
type CancelEmailChangeResponse struct {
	Message string `json:"message"`
}

// ResendVerificationCodeRequest запрос на повторную отправку кода
type ResendVerificationCodeRequest struct {
	RequestID string `json:"request_id" validate:"required"`
}

// ResendVerificationCodeResponse ответ на повторную отправку кода
type ResendVerificationCodeResponse struct {
	Message string `json:"message"`
}

// EmailChangeRequestResponse информация о запросе на смену email
type EmailChangeRequestResponse struct {
	ID        string    `json:"id"`
	OldEmail  string    `json:"old_email"`
	NewEmail  string    `json:"new_email"`
	Status    string    `json:"status"`
	ExpiresAt time.Time `json:"expires_at"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// EmailChangeAuditLogResponse запись аудит-лога
type EmailChangeAuditLogResponse struct {
	ID         string    `json:"id"`
	OldEmail   string    `json:"old_email"`
	NewEmail   string    `json:"new_email"`
	ChangeType string    `json:"change_type"`
	IPAddress  *string   `json:"ip_address,omitempty"`
	UserAgent  *string   `json:"user_agent,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
}

// EmailChangeStatusResponse статус запроса на смену email
type EmailChangeStatusResponse struct {
	HasActiveRequest bool                        `json:"has_active_request"`
	Request          *EmailChangeRequestResponse `json:"request,omitempty"`
}

// EmailChangeAuditLogsResponse список записей аудит-лога
type EmailChangeAuditLogsResponse struct {
	Logs  []EmailChangeAuditLogResponse `json:"logs"`
	Total int                           `json:"total"`
}

